node/src/components/chain_synchronizer/operations.rs

```diff
@@ -552,141 +507,203 @@ async fn sync_trie_store(
+            _ => {}
+        }
+
+        if let Some(key_block_info) = KeyBlockInfo::maybe_from_block_header(
+            &current_header_to_walk_back_from,
+            ctx.config.verifiable_chunked_hash_activation(),
+        ) {
+            check_block_version(ctx.trusted_block_header, ctx.config.protocol_version())?;
+            break Ok(key_block_info);
+        }
+
+        if current_header_to_walk_back_from.height() == 0 {
+            break Err(Error::HitGenesisBlockTryingToGetTrustedEraValidators {
+                trusted_header: ctx.trusted_block_header.clone(),
+            });
+        }
+
+        current_header_to_walk_back_from = *fetch_and_store_block_header(
+            *ctx.effect_builder,
+            ctx.config,
+            *current_header_to_walk_back_from.parent_hash(),
+        )
+        .await?;
+    };
     debug!("finish - fetch the last switch block before the trusted block - fast sync");
+    result
+}
 
-    // Get the most recent header which has the same version as ours
-    // We keep fetching by height until none of our peers have a block at that height and we are in
-    // the current era.
-    let mut most_recent_block_header = trusted_block_header;
+/// Get the most recent header which has the same version as ours
+/// We keep fetching by height until none of our peers have a block at that height and we are in
+/// the current era.
+async fn fetch_block_headers_up_to_the_most_recent_one(
+    trusted_key_block_info: &KeyBlockInfo,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<(BlockHeader, KeyBlockInfo), Error> {
+    let mut most_recent_block_header = ctx.trusted_block_header.clone();
+    let mut most_recent_key_block_info = trusted_key_block_info.clone();
     info!("start - fetch block headers up to the most recent one - fast sync");
     loop {
         let maybe_fetched_block = fetch_and_store_next::<BlockHeaderWithMetadata>(
-            effect_builder,
-            config,
             &most_recent_block_header,
-            &trusted_key_block_info,
+            &most_recent_key_block_info,
+            ctx,
         )
         .await?;
-        match maybe_fetched_block {
-            Some(more_recent_block_header_with_metadata) => {
-                most_recent_block_header = more_recent_block_header_with_metadata.block_header;
-                // If the new block is a switch block, update the validator weights, etc...
-                if let Some(key_block_info) = KeyBlockInfo::maybe_from_block_header(
-                    &most_recent_block_header,
-                    config.verifiable_chunked_hash_activation(),
-                ) {
-                    trusted_key_block_info = key_block_info;
-                }
+
+        if let Some(more_recent_block_header_with_metadata) = maybe_fetched_block {
+            most_recent_block_header = more_recent_block_header_with_metadata.block_header;
+
+            // TODO: If the new block is a switch block, update the validator weights, etc...
```

Comment:
That's done, isn't it?

```suggestion
            // If the new block is a switch block, update the validator weights, etc...
```

Reply:
I see now - if the header contains the "next era validator weights" it means that it's a switch block.
So yes, it's done, indeed.

https://github.com/casper-network/casper-node/pull/2664/commits/1499d9b250d6b4eb5c91f91dd6f30bb4aa6e431e